seqr/views/apis/report_api.py

```diff
@@ -893,6 +932,112 @@ def _get_experiment_ids(airtable_sample, airtable_metadata):
+                f'The following columns are included in the "{file_name}" table but are missing from the data model: {col_summary}'
+            )
+        missing_columns = set(table_validator.keys()).difference(columns)
+        if missing_columns:
+            col_summary = ', '.join(sorted(missing_columns))
+            warnings.append(
+                f'The following columns are included in the "{file_name}" data model but are missing in the report: {col_summary}'
+            )
+
+        for column in columns:
+            _validate_column_data(
+                column, file_name, data, column_validator=table_validator.get(column, {}),
+                warnings=warnings, errors=errors,
+            )
+
+    if errors:
+        raise ErrorsWarningsException(errors, warnings)
+
+    return files, warnings
+
+
+def _load_data_model_validators():
+    response = requests.get(GREGOR_DATA_MODEL_URL)
+    response.raise_for_status()
+    table_models = response.json()['tables']
+    validators = {
+        t['table']: {c['column']: c for c in t['columns']}
+        for t in table_models
+    }
+    required_tables = {t['table'] for t in table_models if t.get('required')}
+    return validators, required_tables
+
+
+def _validate_column_data(column, file_name, data, column_validator, warnings, errors):
+    enum = column_validator.get('enumerations')
+    required = column_validator.get('required')
+    recommended = column in WARN_MISSING_TABLE_COLUMNS.get(file_name, [])
+    if not (required or enum or recommended):
+        return
+
+    missing = []
+    warn_missing = []
+    invalid = []
+    for row in data:
+        value = row.get(column)
+        if not value:
+            if required:
+                missing.append(_get_row_id(row))
+            elif recommended:
+                warn_missing.append(_get_row_id(row))
+        elif enum and value not in enum:
+            invalid.append(f'{_get_row_id(row)} ({value})')
+    if missing or warn_missing or invalid:
+        airtable_summary = ' (from Airtable)' if column in ALL_AIRTABLE_COLUMNS else ''
+        error_template = f'The following entries {{issue}} "{column}"{airtable_summary} in the "{file_name}" table'
+        if missing:
+            errors.append(
+                f'{error_template.format(issue="are missing required")}: {", ".join(sorted(missing))}'
+            )
+        if invalid:
+            invalid_values = f'Invalid values: {", ".join(sorted(invalid))}'
+            errors.append(
+                f'{error_template.format(issue="have invalid values for")}. Allowed values: {", ".join(enum)}. {invalid_values}'
+            )
+        if warn_missing:
+            warnings.append(
+                f'{error_template.format(issue="are missing recommended")}: {", ".join(sorted(warn_missing))}'
+            )
+
+
+def _get_row_id(row):
+    id_col = next(col for col in ['participant_id', 'experiment_sample_id', 'family_id'] if col in row)
+    return row[id_col]
```

Comment:
It may be better to return a row number when it doesn't have an ID column. But I wonder if it is necessary.